physics/sfc_land.f

```diff
@@ -0,0 +1,154 @@
+!            sncovr1, qsurf, evap, hflx, ep, t2mmp, q2mp,               !
+!            gflux, runoff, drain, cmm, chh, zvfun,                     !
+!            errmsg, errflg)                                            !
+!                                                                       !
+!  ====================  defination of variables  ====================  !
+!                                                                       !
+!  inputs:
+!     im          - integer, horiz dimension
+!     cpllnd      - logical, flag for land coupling
+!     cpllnd2atm  - logical, flag for land coupling (lnd->atm)
+!     flag_iter   - logical, flag for iteration
+!     dry         - logical, eq T if a point with any land 
+!     sncovr1_lnd - real   , surface snow area fraction 
+!     qsurf_lnd   - real   , specific humidity at sfc      
+!     evap_lnd    - real   , evaporation from latent heat
+!     hflx_lnd    - real   , sensible heat
+!     ep_lnd      - real   , surface upward potential latent heat flux 
+!     t2mmp_lnd   - real   , 2m temperature 
+!     q2mp_lnd    - real   , 2m specific humidity
+!     gflux_lnd   - real   , soil heat flux over land
+!     runoff_lnd  - real   , surface runoff
+!     drain_lnd   - real   , subsurface runoff  
+!     cmm_lnd     - real   , surface drag wind speed for momentum
+!     chh_lnd     - real   , surface drag mass flux for heat and moisture
+!     zvfun_lnd   - real   , function of surface roughness length and green vegetation fraction
+!  outputs:
+!     sncovr1     - real   , snow cover over land
+!     qsurf       - real   , specific humidity at sfc
+!     evap        - real   , evaporation from latent heat
+!     hflx        - real   , sensible heat
+!     ep          - real   , potential evaporation 
+!     t2mmp       - real   , temperature at 2m 
+!     q2mp        - real   , specific humidity at 2m
+!     gflux       - real   , soil heat flux over land
+!     runoff      - real   , surface runoff
+!     drain       - real   , subsurface runoff
+!     cmm         - real   , surface drag wind speed for momentum
+!     chh         - real   , surface drag mass flux for heat and moisture
+!     zvfun       - real   , function of surface roughness length and green vegetation fraction
+!  ====================    end of description    =====================  !
+!
+!
+      use machine , only : kind_phys
+      implicit none
+
+!  ---  inputs:
+      integer, intent(in) :: im
+      logical, intent(in) :: cpllnd, cpllnd2atm
+      logical, dimension(:), intent(in) :: flag_iter
+      logical, dimension(:), intent(in) :: dry
+
+      real (kind=kind_phys), dimension(:), intent(in) ::                &
+     &       sncovr1_lnd, qsurf_lnd, evap_lnd, hflx_lnd, ep_lnd,        &
+     &       t2mmp_lnd, q2mp_lnd, gflux_lnd, runoff_lnd, drain_lnd,     &
+     &       cmm_lnd, chh_lnd, zvfun_lnd
+
+!  ---  outputs:
+      real (kind=kind_phys), dimension(:), intent(out) ::               &
```

Comment:
Since these are `intent(out)` they will be re-initialized with garbage values. If `cpllnd2atm` if false, then this routine will effectively return these garbage values. I don't know if that matters (if these variables are or ever will be used in the suite after this point when `cpllnd2atm` is false), but it might be safer to have these as `intent(inout)` so that they are not re-initialized.

Reply:
There is a CCPP rule that all `intent(out)` variables are initialized within the routine, even if it returns immediately. So, I think that the choices are to initialize them before returning (when `cpllnd2atm` if false) or change them to `intent(inout)`.

Reply:
@grantfirl These variables are initialized under FV3/ccpp/data/GFS_typedefs.F90. So, I think it would be safe to change them to `intent(inout)`. Do you think that it would be safe to go with it?
Maybe this could be a discussion point in the CCPP Visioning Workshop and I think CCPP need to enforce those kind of standards and conventions in someway. Otherwise, it would be hard to keep the physics code in a certain quality.

Reply:
Ya, I think switching to `intent(inout)` is fine. The documentation (https://ccpp-techdoc.readthedocs.io/en/latest/CompliantPhysicsParams.html#coding-rules) mentions that "All intent(out) variables must be set inside the subroutine, including the mandatory variables errflg and errmsg.", but there is no automatic code-checking to enforce this. We've been relying on code reviews for this, but I agree, the gold standard would be to have the CCPP framework use a Fortran parser to check this.